File: trunk/hangman_cpp/MyGuessingStrategy.cpp
#include <set>
#include <algorithm>
#include <assert.h>

#include "MyGuessingStrategy.h"
#include "HangmanGame.h"
#include "Guess.h"
#include "GuessLetter.h"
#include "GuessWord.h"

////////////////////////////////////////////////////////////////////////////////
//
//  Class MyGuessingStrategy::WordSet
//

/**
 * Sort letters in a word set on their frquencies in descending order.
 */
void MyGuessingStrategy::WordSet::makeOrder() {
  assert(m_order.empty());

  vector<Pair> values(m_stat.begin(), m_stat.end());
  sort(values.begin(), values.end(), LessThan());
  m_order.resize(values.size());
  for (int i = 0; i < values.size(); i++) {
    m_order[i] = values[i].first;
  }
}

/**
 * Insert a new word to the word set, and update the letter statistical info.
 * Letters in the excluded set are not counted for the statistical info.
 */
void MyGuessingStrategy::WordSet::update(const string & word, const set<char> * excluded) {
  set<char> parsed;
  unsigned int i;
  for (i = 0; i < word.length(); i++) {
    char ch = word[i];
    if (!excluded || !excluded->count(ch)) {
      LetterStat & stat = m_stat[ch];
      stat.m_frequency++;
      if (parsed.insert(ch).second) {
        stat.m_wordCount++;
      }
    }
  }
  m_words.push_back(word);
}

/**
 * Give a suggest of the most probable letter not in the excluded letter set.
 * When given a pos, start searching m_order from it and update it to point to a new
 * starting index for the next call.
 */
char MyGuessingStrategy::WordSet::suggest(const set<char> & excluded, unsigned int * pos) {
  if (m_order.empty()) {
    makeOrder();
  }
  unsigned int i = pos ? *pos : 0;
  for (; i < m_order.size(); i++) {
    if (!excluded.count(m_order[i])) {
      if (pos)
        *pos = i + 1;
      return m_order[i];
    }
  }
  assert(0);
  return '\0';
}

////////////////////////////////////////////////////////////////////////////////
//
//  Helper functions used in MyGuessingStrategy
//

/**
 * Return true only when str matches pattern EXACTLY. e.g. given a pattern "AB-",
 * string "ABC" and "ABD" match it, WHILE "ABA" "ABB" AND "XYZ" DON'T. In the same
 * example, the patternChars argument must be {'A', 'B'}
 */
static bool match(const string & pattern, const string & word, const set<char> patternChars) {
  assert(pattern.size() == word.size());
  int size = pattern.size();
  bool ret = true;
  for (int i = 0; i < size; i++) {
    if (pattern[i] != HangmanGame::MYSTERY_LETTER) {
      if (pattern[i] != word[i]) {
        ret = false;
        break;
      }
    }
    else {
      if (patternChars.count(word[i])) {
        ret = false;
        break;
      }
    }
  }
  return ret;
}

/**
 * Fill the first '-' character in pattern with ch.
 * Return a new string of the filled version.
 */
static inline string fillBlank(const string & pattern, char ch) {
  string word = pattern;
  string::size_type idx = word.find(HangmanGame::MYSTERY_LETTER);
  assert(idx != string::npos);
  word[idx] = ch;
  return word;
}

/**
 * If a word has any characters in chars, return true, otherwise false.
 */
static inline bool hasAny(const string & word, const set<char> & chars) {
  for (unsigned int i = 0; i < word.length(); i++) {
    if (chars.count(word[i]) > 0)
      return true;
  }
  return false;
}

////////////////////////////////////////////////////////////////////////////////
//
//  Class MyGuessingStrategy
//

/**
 * Initialize with a dictionary of words.
 */
MyGuessingStrategy::MyGuessingStrategy(const set<string> & dict) {
  set<string>::const_iterator it = dict.begin();
  set<string>::const_iterator end = dict.end();
  vector<string> patterns;
  for (; it != end; ++it) {
    const string & word = toUpperCase(*it);
    unsigned int len = word.length();
    if (patterns.size() < len)
      patterns.resize(len);
    string & pattern = patterns[len - 1];
    if (pattern.empty())
      pattern.assign(len, HangmanGame::MYSTERY_LETTER);
    if (m_patternMapGroup.size() < len)
      m_patternMapGroup.resize(len);
    m_patternMapGroup[len - 1][pattern].update(word);
  }
}

/**
 * Make a new WordSet to the pattern.
 */
MyGuessingStrategy::WordSet & MyGuessingStrategy::newWordSet(const string & pattern) {
  int len = pattern.length();
  PatternMap & map = m_patternMapGroup[len - 1];

  //Find the smallest "parent" pattern collection
  PatternMap::iterator end = map.end();
  PatternMap::iterator found = end;
  set<char> patternChars; //e.g. given a pattern "AB-A--", patternChars is {'A', 'B'}
  for (int i = 0; i < len; i++) {
    char ch = pattern[i];
    if (ch != HangmanGame::MYSTERY_LETTER && patternChars.insert(ch).second) {
      string copy = pattern;
      copy[i] = HangmanGame::MYSTERY_LETTER;
      for (int j = i + 1; j < len; j++) {
        if (copy[j] == ch)
          copy[j] = HangmanGame::MYSTERY_LETTER;
      }

      PatternMap::iterator iter = map.find(copy);
      if (iter != end && (found == end || found->second.wordCount() > iter->second.wordCount())) {
        found = iter;
      }
    }
  }
  assert(found != end);

  //Draw the new pattern collection and info through filtering the parent
  WordSet & parentSet = map[found->first];
  WordSet & newSet = map[pattern];

  vector<string>::const_iterator iter = parentSet.words().begin();
  vector<string>::const_iterator tail = parentSet.words().end();
  for(; iter != tail; ++iter) {
    const string & word = *iter;
    if (match(pattern, word, patternChars)) {
      newSet.update(word, &patternChars);
    }
  }
  return newSet;
}

/**
 * When we have a last chance to make a guess and there're more than one blanks
 * in a pattern, we do the final blow! The basic idea is to select a word, which
 * dosn't contain any wrong guessed letters while having the most probable letter
 * given by WordSet::suggest.
 */
string MyGuessingStrategy::finalBlow(const string & pattern, WordSet & wordset, const set<char> & wrongLetters) {
  string guess;
  vector<string> candidates;
  unsigned int i = 0;
  char ch = wordset.suggest(wrongLetters, &i);
  vector<string>::const_iterator iter = wordset.words().begin();
  vector<string>::const_iterator tail = wordset.words().end();
  for(; iter != tail; ++iter) {
    const string & word = *iter;
    if (!hasAny(word, wrongLetters)) {
      candidates.push_back(word);
      if (word.find(ch) != string::npos) {
        guess = word;
        break;
      }
    }
  }

  while (guess.empty()) {
    unsigned int j;
    char ch = wordset.suggest(wrongLetters, &i);
    for (j = 0; j < candidates.size(); j++) {
      if (candidates[j].find(ch) != string::npos) {
        guess = candidates[j];
        break;
      }
    }
  }

  return guess;
}

/**
 * Suggest a letter or word.
 */
string MyGuessingStrategy::suggest(const string & pattern, WordSet & wordset, const HangmanGame & game) {
  //If the pattern collection has only one word, that's it!
  if (wordset.wordCount() == 1) {
    return wordset.words()[0];
  }

  //Make a guess, according to letter frequency of a pattern.
  string word;
  const set<char> & wrongLetters = game.getIncorrectlyGuessedLetters();
  const set<string> & wrongWords = game.getIncorrectlyGuessedWords();
  //Number of '-' characters in a pattern.
  int patternBlanks = count(pattern.begin(), pattern.end(), HangmanGame::MYSTERY_LETTER);

  if (patternBlanks > 1) {
    if (game.numWrongGuessesRemaining() == 0)
      word = finalBlow(pattern, wordset, wrongLetters);
    else
      word = wordset.suggest(wrongLetters);
  }
  else {
    unsigned int i = 0;
    set<char> excluded = wrongLetters;
    while(1) {
      char ch = wordset.suggest(excluded, &i);
      word = fillBlank(pattern, ch);
      if (!wrongWords.count(word))
        break;
      else
        excluded.insert(ch);
    }
  }

  assert(!word.empty());
  return word;
}

Guess & MyGuessingStrategy::nextGuess(const HangmanGame & game) {
  const string & pattern = game.getGuessedSoFar(); //This line clearly explains what a "pattern" is.
  WordSet & wordset = wordSet(pattern);
  string next = suggest(pattern, wordset, game);
  if (next.length() == 1) {
    m_guessLetter.setLetter(next[0]);
    return m_guessLetter;
  }
  else {
    m_guessWord.setWord(next);
    return m_guessWord;
  }
}

File: hangman_java/MyGuessingStrategy.java
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.Map;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Collection;
import java.util.AbstractCollection;

class MyGuessingStrategy implements GuessingStrategy {
  /**
   * A WordSet is a set of words having the same pattern.
   * A pattern is a string returned by HangmanGame::getGuessedSoFar. It's like
   * "AB-", which matches words like "ABC", "ABD" and "ABX", but not "ABA" or "ABB".
   * A WordSet contains statistical info about letters that are NOT GUESSED yet.
   */
  private static class WordSet extends AbstractCollection<String> {

    private class WordIterator implements Iterator<String> {
      private Iterator<String> it = WordSet.this.words.iterator();

      @Override
      public boolean hasNext() {
        return it.hasNext();
      }

      @Override
      public String next() {
        return it.next();
      }

      @Override
      public void remove() {
        throw new UnsupportedOperationException();
      }
    }

    private static class LetterStat implements Comparable<LetterStat> {
      public final char ch;
      public int count = 0;  //How many times the letter appears in a WordSet
      public int wordCount = 0;  //How many words contains the letter in a WordSet

      LetterStat(char ch) {
        this.ch = ch;
      }

      @Override
      public int compareTo(LetterStat rhs) {
        if (this.count > rhs.count)
          return -1;

        if (this.count == rhs.count)
          return this.wordCount - rhs.wordCount;

        return 1;
      }
    }

    /**
     * A map of letters to their statistical info.
     */
    private Map<Character, LetterStat> stat = new HashMap<Character, LetterStat>();

    /**
     * Letters in descendent order on frequency
     */
    private List<LetterStat> order = null;

    /**
     * Words in the set.
     */
    private List<String> words = new ArrayList<String>();

    public final String pattern;

    public final Set<Character> guessedLetters;

    WordSet(String pattern, Set<Character> guessedLetters, Collection<String> words) {
      assert(pattern != null && guessedLetters != null);
      this.pattern = pattern;
      this.guessedLetters = guessedLetters;
      addAll(words);
    }

    /**
     * Determine if a word matches the pattern and guessed letters.
     */
    private boolean match(String word) {
      int size = pattern.length();
      assert(size == word.length());

      boolean ret = true;
      for (int i = 0; i < size; i++) {
        if (pattern.charAt(i) != HangmanGame.MYSTERY_LETTER) {
          if (pattern.charAt(i) != word.charAt(i)) {
            ret = false;
            break;
          }
        }
        else {
          if (guessedLetters.contains(word.charAt(i))) {
            ret = false;
            break;
          }
        }
      }
      return ret;
    }

    @Override
    public boolean add(String word) {
      if (!match(word)) {
        //throw new IllegalArgumentException();
        //It really should raise the exception when the word is not acceptable.
        //However, returning false can make addAll work conveniently on a
        //collection of words.
        return false;
      }

      Set<Character> parsed = new HashSet<Character>();
      for (int i = 0; i < word.length(); i++) {
        char ch = word.charAt(i);
        if (!guessedLetters.contains(ch)) {
          LetterStat stat = this.stat.get(ch);
          if (stat != null) {
            stat.count++;
            if (parsed.add(ch)) {
              stat.wordCount++;
            }
          }
          else {
            stat = new LetterStat(ch);
            stat.count++;
            stat.wordCount++;
            this.stat.put(ch, stat);
            parsed.add(ch);
          }
        }
      }
      this.words.add(word);
      return true;
    }

    @Override
    public int size() {
      return words.size();
    }

    @Override
    public Iterator<String> iterator() {
      return new WordIterator();
    }

    /**
     * Give a suggest of the most probable letter not in the excluded letter set.
     */
    public char suggest(Set<Character> excluded) {
      if (this.order == null) {
        makeOrder();
      }
      for (int i = 0; i < this.order.size(); i++) {
        char ch = this.order.get(i).ch;
        if (!excluded.contains(ch)) {
          return ch;
        }
      }
      assert(false);
      return '\0';
    }

    /**
     * Sort letters in a word set on their frquencies in descending order.
     */
    private void makeOrder() {
      order = new ArrayList<LetterStat>(this.stat.values());
      Collections.sort(order);
    }

    @Override
    public String toString() {
      return "WordSet[" + size() + "]";
    }
  };

  private WordSet wordset;

  public MyGuessingStrategy(HangmanGame game, Set<String> dict) {
    String pattern = game.getGuessedSoFar();
    int len = pattern.length();
    List<String> words = new ArrayList<String>();
    for (String word: dict) {
      if (word.length() == len) {
        words.add(word);
      }
    }
    this.wordset = new WordSet(pattern, game.getAllGuessedLetters(), words);
  }

  @Override
  public Guess nextGuess(HangmanGame game) {
    //Update wordset on a previous guess result, be it successful or not.
    this.wordset = new WordSet(game.getGuessedSoFar(), game.getAllGuessedLetters(), this.wordset);
    String next = suggest(game);
    if (next.length() == 1) {
      return new GuessLetter(next.charAt(0));
    }
    else {
      return new GuessWord(next);
    }
  }

  /**
   * When we have a last chance to make a guess and there're more than one blanks
   * in a pattern, we do the final blow! The basic idea is to select a word, which
   * dosn't contain those wrong guessed letters while has the most probable
   * letter given by a WordSet::suggest.
   */
  private String finalBlow(HangmanGame game) {
    String guess = null;
    List<String> candidates = new ArrayList<String>();
    char ch = this.wordset.suggest(game.getAllGuessedLetters());
    Set<Character> wrongLetters = game.getIncorrectlyGuessedLetters();

    for (String word : this.wordset) {
      if (!hasAny(word, wrongLetters)) {
        candidates.add(word);
        if (word.indexOf(ch) != -1) {
          guess = word;
          break;
        }
      }
    }

    if (guess == null) {
      Set<Character> excluded = new HashSet<Character>(wrongLetters);
      while (guess == null) {
        excluded.add(ch);
        ch = this.wordset.suggest(excluded);
        for (String word : candidates) {
          if (word.indexOf(ch) != -1) {
            guess = word;
            break;
          }
        }
      }
    }

    assert(guess != null);
    return guess;
  }

  /**
   * Suggest a letter or word.
   */
  private String suggest(HangmanGame game) {
    if (this.wordset.size() == 1) {
      return this.wordset.iterator().next();
    }

    String word = null;
    Set<Character> guessedLetters = game.getAllGuessedLetters();
    String pattern = game.getGuessedSoFar();
    int patternBlanks = numOfBlanks(pattern); //Number of '-' characters in a pattern.

    if (patternBlanks > 1) {
      if (game.numWrongGuessesRemaining() == 0) {
        word = finalBlow(game);
      }
      else {
        word = Character.toString(this.wordset.suggest(guessedLetters));
      }
    }
    else {
      //When there's only one blank letter, try to guess the word to save one
      //score on a successfull guess.
      Set<Character> excluded;
      Set<String> wrongWords = game.getIncorrectlyGuessedWords();
      if (!wrongWords.isEmpty()) {
        //Reduce the wordset by incorrectly guessed letters from the wrongWords
        //and guessedLetters.
        int idx = pattern.indexOf(HangmanGame.MYSTERY_LETTER);
        excluded = new HashSet<Character>(guessedLetters);
        for (String wd : wrongWords) {
          excluded.add(wd.charAt(idx));
        }
        this.wordset = new WordSet(pattern, excluded, this.wordset);
      }
      else {
        excluded = guessedLetters;
      }
      char ch = this.wordset.suggest(excluded);
      word = pattern.replace(HangmanGame.MYSTERY_LETTER, ch);
    }

    assert(word != null);
    return word;
  }

  /**
   * If a word has any characters in chars, return true, otherwise false.
   */
  private static boolean hasAny(String word, Set<Character> chars) {
    for (int i = 0; i < word.length(); i++) {
      if (chars.contains(word.charAt(i)))
        return true;
    }
    return false;
  }

  private static int numOfBlanks(String pattern) {
    int count = 0;
    for (int i = 0; i < pattern.length(); i++) {
      if (pattern.charAt(i) == HangmanGame.MYSTERY_LETTER)
        count++;
    }
    return count;
  }

  @Override
  public String toString() {
    return "MyGuessingStrategy[" + this.wordset.toString() + "]";
  }
}
